### DISCUSSION TITLE
In Docker, migrating an existing intermediate

### DISCUSSION THREAD

**Discussion by MP-ICT-Diensten** [smallstep/certificates#995] [Q&A] [answered]
### Steps to Reproduce

Hello,

I installed step-ca in a dev environment in docker and followed the manual from here (The medium way):
https://smallstep.com/docs/tutorials/intermediate-ca-new-ca

After replacing the root ca, intermediate ca and intermediate key files and starting step-ca again I have the following error:
`step-ca  | error decrypting /home/step/secrets/tts.local_intermed-ca.key: **x509: decryption password incorrect**`

What I do see is that the certificates generated by step-ca are much smaller in size:

bash-5.1$ ls -al
total 16
drwx--S---    2 step     step           114 Aug  3 14:39 .
drwxr-sr-x    7 step     step            96 Aug  3 14:22 ..
-rw-------    1 step     step           725 Jul 27 15:49 intermediate_ca.crt
-rw-------    1 step     step           668 Jul 27 15:49 root_ca.crt
-rw-------    1 step     step          2179 Aug  3 15:07 tts.local_intermed-ca.crt
-rw-------    1 step     step          2439 Aug  3 14:47 tts.local_root-ca.crt

I wonder what is going wrong here.

### Your Environment

* OS - Oracle Linux 8.5
* `step-ca` Version -  docker:  smallstep/step-ca   0.21.0  

### Expected Behavior

After replacing the files in the ca.json the intermediat key should be read correctly and be used.

### Actual Behavior

Intermediate key could not be used.

### Additional Context

Difference in size of default step-ca generated and our (development testing) intermediate certificate.

bash-5.1$ more intermediate_ca.crt
-----BEGIN CERTIFICATE-----
MIIB6TCCAY+gAwIBAgIRAKt2gLc5eYILnyDNGnbIV2swCgYIKoZIzj0EAwIwPjEZ
MBcGA1UEChMQVFRTLVRlc3RvbWdldmluZzEhMB8GA1UEAxMYVFRTLVRlc3RvbWdl
dmluZyBSb290IENBMB4XDTIyMDcyNzE1NDk0MVoXDTMyMDcyNDE1NDk0MVowRjEZ
MBcGA1UEChMQVFRTLVRlc3RvbWdldmluZzEpMCcGA1UEAxMgVFRTLVRlc3RvbWdl
dmluZyBJbnRlcm1lZGlhdGUgQ0EwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAAQP
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
FoAU1WFIAG/4lek25WtSnkYYJtcH79UwCgYIKoZIzj0EAwIDSAAwRQIgEgKtmKkf
HkSevAE3xhOjndN14uDfr7UaKDBbbu2OVN8CIQDpHGcgi0wcNVOIbkyvNhrdZC2c
JoAp4nPp7Xy5si+yNA==
-----END CERTIFICATE-----
bash-5.1$
bash-5.1$
bash-5.1$ more tts.local_intermed-ca.crt
-----BEGIN CERTIFICATE-----
MIIGGzCCBAOgAwIBAgIQGnV4+xzWrF7W7rkNYfeEqTANBgkqhkiG9w0BAQsFADA9
MRIwEAYDVQQKDAl0dHMubG9jYWwxJzAlBgNVBAMMHnR0cy5sb2NhbCBSQ0EgVFRT
IFRlc3RvbWdldmluZzAeFw0yMjA1MDMwMDAwMDBaFw0zMjA1MDUwMDAwMDBaMD4x
EjAQBgNVBAoMCXR0cy5sb2NhbDEoMCYGA1UEAwwfdHRzLmxvY2FsIEktQ0EgVFRT
IFRlc3RvbWdldmluZzCCAaIwDQYJKoZIhvcNAQEBBQADggGPADCCAYoCggGBALUJ
/j8MyfbjYAjv896JOShVz27Ql0ACP+HpLpMWf6jh2dQwcMLaiMpJmrDJDI7cYKoa
4LKUFT2ww57oPw5w5StPB7M+VJtLBlctzCepkKCirTKT3cRt27zGKc/seITN5/Hc
C+gGUKnSIA4lwN/w4FTrC32AV6vb48B/3jBW+wixsX+RbZp7P+LtPuzGs5i232n7
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
YxlSFCTpX63F30wUefhsrBbrtPCe/HNEh88gKIizw0fOy8pNiDmV4G6XuIi4k4el
tN+HyTMoLCI81BSS2BlY2ZEOnWYe8o9GU0/fyJs75IeiqFvqmE2Y1N9rIpPd6wID
AQABo4IBlDCCAZAwEgYDVR0TAQH/BAgwBgEB/wIBADAOBgNVHQ8BAf8EBAMCAQYw
HQYDVR0OBBYEFH5jURx8PrmdKkCVDynpE4heSEGBMDYGA1UdEQQvMC2GFGh0dHA6
Ly9jYS50dHMubG9jYWwvgRVMMi1MMy1CZWhlZXJAcmNzbmwuZXUwHwYDVR0jBBgw
FoAUmaT3m8GpKAQNG2BnHNSJUzVIR70wNgYDVR0SBC8wLYYUaHR0cDovL2NhLnR0
cy5sb2NhbC+BFUwyLUwzLUJlaGVlckByY3NubC5ldTBlBggrBgEFBQcBAQRZMFcw
VQYIKwYBBQUHMAKGSWh0dHA6Ly9jYS50dHMubG9jYWwvY2VydHMvdHRzLmxvY2Fs
X1Jvb3RfQ2VydGlmaWNhdGlvbl9BdXRob3JpdHkuY2VydC5wZW0wUwYDVR0fBEww
SjBIoEagRIZCaHR0cDovL2NhLnR0cy5sb2NhbC9jcmwvdHRzLmxvY2FsX1Jvb3Rf
Q2VydGlmaWNhdGlvbl9BdXRob3JpdHkuY3JsMA0GCSqGSIb3DQEBCwUAA4ICAQDk
A4Y09S4fO2klmPDAsXVhAAy4Nf88d1+Aj92Lr/jVGOwk464Pu5R/VK0m+tcfHXNu
OIjo52rWTJiVj1mFksajT7fC0FlKgabda8w5+wNT6eQeyqxq3F2qWR0XeYixDDDi
2lUrS8vHQrQYAdbaG1PkiKsDOmNsOPzTsWi2Lh+KiUJUMBf3u57TVLx2oLXtx4Z0
/fa+2rVzTwDRIq4Xb5M5wyp9rJUtlJ6NFrygdLQHizAUEG71JXsLzYiwAwxYP0N5
I0FmLtJueDKssvDatVxkF9KvKQQOg3g6LC5RMzgPblvllX7X3YYLsoPaMQPMriE4
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
im/ST5d/Jz+foxRvKdcNUACIefC1iBYQbKlHStWlqwidso4jcC3zIsSsgAxGZmNJ
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
9x4/dlA2fR/TBP+SXTyssQx4YbU9EDsKKj2DerwhkTYYgofvqKnqzgji53sbV323
sMUihHhIQGtar0ALv/dtRtXGy/Wyu5vU87JeGfHXrw==
-----END CERTIFICATE-----

### Contributing

Vote on this issue by adding a 👍 reaction.
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by tashian** [comment, marked as answer]
Hi @MP-ICT-Diensten , there's a `password` file (I think `/home/step/password`?) that gets generated on the container volume when the CA is first initialized. You'll need to change the contents of that file, so that it contains the password for your intermediate signing key.

**Reply by MP-ICT-Diensten** [reply]
Yes, thanks, that is the right answer, step-ca now starts successful.  
For me this is a "bug" in the manual, this part about the password file is missing in the[ Medium way ](https://smallstep.com/docs/tutorials/intermediate-ca-new-ca#the-medium-way)procedure.

**Reply by tashian** [reply]
Thanks for the note, we'll take a look at how we might make this clearer in the tutorial.

**Comment by MP-ICT-Diensten** [comment]
Next step, I get now an error in the logs, which returns regularly:

/usr/local/go/src/net/http/server.go:3195: http: TLS handshake error from 127.0.0.1:47934: remote error: tls: bad certificate


To be sure I also checked the fingerprint and adjusted this accordingly in de default.json using the below command:
step certificate fingerprint /home/step/certs/root_ca.crt

**Comment by MP-ICT-Diensten** [comment]
And I have this error when I try any step command:

# step ca policy authority x509 allow dns 'localhost'
No admin credentials found. You must login to execute admin commands.
✔ Please enter admin name/subject (e.g., name@example.com): admin@localhost█
error creating admin client: client GET https://tts-ss-ca.tts.local:9000/provisioners?limit=100 failed: Get "https://tts-ss-ca.tts.local:9000/provisioners?limit=100": x509: a root or intermediate certificate is not authorized to sign for this name: DNS name "localhost" is not permitted by any constraint


Requesting the provisionerslist does work:
# curl https://localhost:9000/provisioners
{"provisioners":[{"type":"JWK","name":"tts-recovery","key":{"use":"sig","kty":"EC","kid":"WYYCva42JO_4Vnuxp-WsuImtzJ37lq7ofjI0TDqJSCU","crv":"P-256","alg":"ES256","x":"4IKxF_zLcdgvBCu_FXPxovSVB-Or4KDYjhZdTTvvS-o","y":"rjBTlRoaxDkQmOJVZ645JeMPOB3OmUwPncnWiEww2Mo"},"encryptedKey":"eyJhbGciOiJQQkVTMi1IUzI1NitBMTI4S1ciLCJjdHkiOiJqd2sranNvbiIsImVuYyI6IkEyNTZHQ00iLCJwMmMiOXXXXXXXXXXXXXXXXDJzIjoiS29FV3pCUFktSVZnNkFHXzJCSklrZyJ9.J9zsGfVFfN9k1zhw8k0b9mnyOwCm4LBbALW6sKMycXwgwsKKuvPRlA.AkOOOpRd5H2Jlgyh.7JWb89YkUSvz8JIpdmzWeaodRa1LuVfQtEeXIZxd_JXuOuRL3n2gHDBQoboYy5kTw-V-eWvXIXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX_F-BtC0GX3W4jGwBOI-5UiC7_gma8B4n6T5Zfb5ghj7bii5sfuocmIUZQuL_pryrG9NzPYDPBk3z1oD2PuDX630Lx2h7Gn_1Azekxt7eA2Llv9qJNAN6mru8viIhcY2k0xTyj3fT6zk1Q8oDtb42fzYD8yfWfT6at3z7N7qh9LhPuE3r_R1MevjtlpQE7SnEa8wqjN-tmJ-rjA.Vh1ziy2FtwQQ4Zo1B-yHpA"}],"nextCursor":""}

**Reply by tashian** [reply]
When step-ca starts up, it uses the intermediate cert to sign its own leaf certificate (for CA traffic) that has the "dnsNames" (from config.json) as subjects. What this error sounds like, to me, is that you have "localhost" in your DNS names config, but you have a root or intermediate cert with X.509 Name Constraints on it, that are preventing the CA from signing its own leaf certificate.

**Reply by tashian** [reply]
Can you tell me a bit more about the CA software you are migrating from, that generated your intermediate CA?
Just to set expectations here, it's possible that your CA certs are not going to be compatible with`step-ca`, because there are many possible migration scenarios and we don't support all of them.
But, let's see how far we can get with it.

If you can share the output of `step certificate inspect` on your intermediate (with redactions as needed), it may help!

**Reply by MP-ICT-Diensten** [reply]
The constrains are in the root certificate and not in the intermediate:
```
            X509v3 Name Constraints: critical
                Permitted:
                  DNS: .tts.local, .***nl.eu
                  Email: tts.local, ***nl.eu
```

But I think I figured it out now,:

- Install docker and make sure that in the ca.json the hostnames and ip adresses are used which are within the constrians.
- Then replace the root and intermediate certificates
- Then replace intermediat key file and delete old root ca file
- Then change the passowrd to the intermediat password
- Generate a new fingerprint based on the rootca and place that in the default.json.
- Restart step-ca docker  
Then in the host with step-cli
- Install root ca chain into host certificate store 
```#On the host you can perform this to install certificate (if not sudo, asked for sudo parrword to install):
{
  CA_FINGERPRINT=$(docker run --rm -v step-ca:/home/step smallstep/step-ca:0.21.0 step certificate fingerprint certs/root_ca.crt)
  step ca bootstrap --ca-url https://ca.tts.local:9000 --fingerprint $CA_FINGERPRINT --install
}
```
- Do a health check and result is ok!

Now in the docker in ca.json is a JMX provisioner, but we don't know the password, thus create a new provisioner and delete the old one and register the password of the new provisioner
```
step ca provisioner list
step ca provisioner remove "old-name"
step ca provisioner add {newname} --type JWK --create    # the --create is missing from web manuals
 Succes!
kill -1 1  # reload step-ca in docker
``` 
And this is fine and I can create a certificate with step ca command.

But, now I am going to try to add a authority and that is failing

```
bash-5.1$ step ca policy authority x509 allow ip 10.253.0.0/16
No admin credentials found. You must login to execute admin commands.
✔ Please enter admin name/subject (e.g., name@example.com): blabla@tts.local
✔ Provisioner: blabla@tts.local (JWK) [kid: dfGJvIpOj9vvCB6gI9u2EWUNB5Keu8wR6eVrFDqVU-A]
Please enter the password to decrypt the provisioner key:
error retrieving policy: json: cannot unmarshal number into Go value of type ca.AdminClientError
bash-5.1$
```

I can manually edit the ca.json file for this, but why is the command not working?  I tried multiple ways to use the cli, but all fail with the same error: **error retrieving policy: json: cannot unmarshal number into Go value of type ca.AdminClientError**

Just a view alos give the same error
````
bash-5.1$ step ca policy provisioner view --provisioner blabla@tts.local
No admin credentials found. You must login to execute admin commands.
✔ Please enter admin name/subject (e.g., name@example.com): blabla@tts.local
✔ Provisioner: blabla@tts.local (JWK) [kid: dfGJvIpOj9vvCB6gI9u2EWUNB5Keu8wR6eVrFDqVU-A]
Please enter the password to decrypt the provisioner key:
error retrieving authority policy: json: cannot unmarshal number into Go value of type ca.AdminClientError
bash-5.1$
```

The provisioner list is there:
```
bash-5.1$ step ca provisioner list

   {
      "type": "JWK",
      "name": "blabla@tts.local",
      "key": {
         "use": "sig",
         "kty": "EC",
         "kid": "dfGJvIpOj9vvCB6gI9u2EWUNB5Keu8wR6eVrFDqVU-A",
         "crv": "P-256",
         "alg": "ES256",
         "x": "4EVbslcJ1OWhFB42MEy-CBBkZ3h1kClMA69JNSPyLsk",
         "y": "-VQrPJzvXskimUTkUsHS9dAFfO-jn3e9MNG6aI1mSEM"
      },
      "encryptedKey": "eyJhbGciOiJQQkVTMi1IUzI1NitBMTI4S1ciLCJjdHkiOiJqd2sranNvbiIsImVuYyI6IkEyNTZHQ00iLCJwMmMiOjEwMDAwMCwicDJzIjoicXJMMkJUTnB0bEVfQ2V3NHJaUUhSUSJ9.EhIq8p876OHt3EBpnn85fU4U0UQubzN3zFPwie6ZPzHFtm_S_FLZ6g.zr_V2Jen-w_lD8Sl.aXUzDlNjdCx8Zi6ND_8cL_GPos7b7Gzq-eYzQwwLc3tD_I6PapYuHbWbZyzzGasGGCrw4c2TK6xZQ2MdEosCzeMQyMs_eyFTVUIdP_EhcX-ibhQ9RIOgz9oRiX1aK-6Cc9DFVKKRqRJaf7V0n1-pApaKBr23OrkbeIdV5TZzQDAgFXPpz113frf8QSm911YGFSII9bqQQ-sj2PnMK9vNbiSEkAGkChEYcanC3F7GuqEF3CLRKFOoJlgwrDv31JuG_dP5LE1gmLrL9VG_gF-GP7NavKchQJO_NbyDGAWUYbJMdXdDYDoBY-UTFaIubn1u7J1Ju02atqqg6mwTwZI.EGq7nt7MoWrsPRhfdnKH5g",
      "claims": {
         "enableSSHCA": true,
         "disableRenewal": false,
         "allowRenewalAfterExpiry": false
      },
      "options": {
         "x509": {},
         "ssh": {}
      }
```
